### DISCUSSION TITLE
Rclone integration

### DISCUSSION THREAD

**Discussion by Tormund1985** [bluenviron/mediamtx#2350] [General]
Hi everyone, 
I have mediamtx configured on a raspberry pi running the recording and streaming of a CSI camera.

On the other hand running through crontab an script that syncs the files created by ffmpeg to a cloud through rclone. However it consumes a lot of resources as it is continuously monitoring the folder.

I wonder if there is a way to improve its efficiency by running a simple upload when ffmpeg finishes writing a file to disk in the mediamtx yml file itself.

How could I make it so that it would run when ffmpeg finishes writing the file?

Thanks in advance.

**Comment by aler9** [comment]
Hello, first of all native recording is being implemented (#2255) and FFmpeg is not necessary anymore. Furthermore, MP4s generated by the native recording system can be read immediately without waiting finalization.

You can use `inotifywait` to launch a command whenever a recording file, generated by either the server of FFmpeg, is finalized and can be safely copied somewhere else:

```
while inotifywait -e close_write /my-recording/directory; do rclone --arg1 --arg2; done
```

**Reply by Tormund1985** [reply]
The native recorder makes much bigger files with ffmpeg (600 vs 60Mb), I don't know how to reduce the size with the native recorder. 

What I would like is that rclone executes at the end of the recording the upload of that file, of course as the file is named with the values %h%m%d etc I don't know how to do it so that it executes only the upload of that file.

**Reply by aler9** [reply]
There are no size differences between files produced by the native recorder and files produced by FFmpeg when there's no re-encoding, i.e. with:

```
ffmpeg -i rtsp://... -c copy myfile.mp4
```

If instead you chose to re-encode the stream, you get smaller files at cost of a lot CPU consumption:

```
ffmpeg -i rtsp://... myfile.mp4
```

Therefore it's a choice you have to take depending on your needs and available computational power.

inotifywait allows you to get the name of the file that has just been created, and you can pass it to rclone:

```sh
inotifywait -q -m -e close_write . | while read -r directory events filename; do rclone $filename other_options; done
```

But, from what i remember, `rclone` detects differences between the local folder and the remote location, and uploads only the new or changed files, so it should be enough to call `rclone sync` without specifying the file name:

```sh
inotifywait -q -m -e close_write . | while read -r directory events filename; do rclone sync . other_options; done
```

**Reply by Tormund1985** [reply]
Sorry but I have to put that command after RunOnReady: my ffmpeg command ; inotifywait -q -m -e close_write . | while read -r directory events filename; do rclone sync "my source and destiny" ; done

Is that correct?

Many thanks

**Reply by aler9** [reply]
in my opinion it's better to run `inotifywait` separately from MediaMTX, since it runs independently from the fact that a source is available or not.

Running two commands in parallel in `runOnDemand` is not trivial, it may be accomplished with:

```yml
paths:
  mypath:
    runOnDemand: sh -c "command1 & command2 &; wait"
```

but i'm not sure, anyway is redundant in this case.

**Reply by Tormund1985** [reply]
Ok, I appreciate your comments, I think I will finally do it with sync, it's a command that works very well with rclone.

And on the other hand you are right, the size of the native recording with respect to ffmpeg is the same, it was different because it had different bitrate. I will check which one consumes more resources as my computers are battery powered so my goal is to save energy as much as possible.

Thank you very much!

[a comment by an automated account is not shown]
